src/rgw/driver/sfs/bucket.cc

```diff
@@ -534,11 +534,39 @@ int SFSBucket::abort_multiparts(
   return sfs::SFSMultipartUploadV2::abort_multiparts(dpp, store, this);
 }
 
+/**
+ * @brief Refresh this bucket's rep with the state obtained from the store.
+ *        Indeed it can happen that the rep of this bucket is obsolete due to
+ *        concurrent threads updating metadata using their own SFSBucket instance.
+ */
 int SFSBucket::try_refresh_info(
```

Comment:
This might be only half the solution. I didn't dig to deep, but hope you know more :)

If I inline and simplify the logic a bit we have: 
```c++
int retry_raced_bucket_write(const DoutPrefixProvider *dpp, rgw::sal::Bucket* b, const F& f) {
  auto r = f();
  for (auto i = 0u; i < 15u && r == -ECANCELED; ++i) {
    r = b->try_refresh_info(dpp, nullptr);
    if (r >= 0) {
      rgw::sal::Attrs attrs = s->bucket->get_attrs();
      attrs[RGW_ATTR_TAGS] = tags_bl;
      r = s->bucket->merge_and_store_attrs(this, attrs, y);
    }
  }
  return r;
}
```

What if something goes state between try_refresh_info and merge_and_store_attrs? Is merge_and_store_attrs responsible to return -ECANCELED when it detects staleness? Can we detect merge conflicts there?

Reply:
If I have to say the truth, I don't think the `retry_raced_bucket_write` can solve the problem for all cases, it can only mitigate the problem.
Imagine 2 **physical** threads A and B and a bucket with an initial state `S0`, they both race and run **true-parallel** on their own CPU core the `try_refresh_info` function; at this point for both A and B the state of the bucket in memory is `S0`, they both think the `S0` state is the latest one and they happily update the bucket both with their new data.
Now, because in any decent implementation, an update will be done mutual exclusively, A or B will run a portion of critical code in a serialized way, so in the end, either the state of A or the state of B will prevaricate on the other.
Is this an error? I don't think so, it is part of the game when you allow a bucket to be modified concurrently by more than 1 physical thread.
At a certain point, a thread will have to say: "Can I safely assume this state in my cached memory is good enough?"; it doesn't matter how much far you go checking this, at a certain point you simply must go on. 
This `retry_raced_bucket_write` mechanism can work well with **logical** threads mapped over 1 **physical** thread.
In this case they can always be serialized in some way; in this case the `retry_raced_bucket_write` solves something it is correct to solve because the 2 logical threads run over a single core as if they where a single logical sequence of statements.
This is the reasoning I've done myself trying to figure out what were the original "meaning and purposes" of the original authors of this code.
Let's always keep in mind that the actual Truth, until we are working downstream, and dealing with parts developed upstream will be hard to be completely clarified; we would need the support of the original developers to really understand what was the aim of something.
The best we can do now is to hope that our guess is good enough :) .

Reply:
Let's assume we are in `s->bucket->merge_and_store_attrs(this, new, y);`

State we have: database state *db*, our current copy in `bucket` *current* and the update in `new`.
Caller based *new* on *current* fetched from *db*.

We must transition from *db* to *new*. Right now we transition form *current* to *new*. This is a bug, because *current* may be stale. This is the case if *current* != *db*.

If we make the update transaction conditional on *current* == *db*, we would transition from *db* to *new* in turn. If at time of the update *current* != *db*, we fail with -ECANCELED and let the retry logic compute a fresh update.

Reply:
Ok, now I think I got what you mean, thanks; I try to explain the limit case so we can discuss from that:
Let's assume both threads A and B reach `update()` concurrently.
Let's assume that `update()` is the function that flushes a state on the db and cannot be executed by 2 threads concurrently because mutex protected.

**thread A**

```
try_update_data()
	-> fetch_from_db() -> db_state_0
	->  current_state = db_state_0

update(current_state + new_A)
```
now bucket's state on db is _db_state_A_

**thread B**

```
try_update_data()
	-> fetch_from_db() -> db_state_0
	-> current_state = db_state_0

update(current_state_0 + new_B)
```
now bucket's state on db is _db_state_B_

**final state on db: _db_state_B_**

**_db_state_A_ != _db_state_B_**

_new_A_ is lost.

To avoid the scenario above,  we should implement a mechanism inside the `update()` mutex protected function (and it must be only that way for the following check).
So `update() `should be something like:

```
int update(const State &current, const &new_delta)
{
  mutex.lock()
  defer mutex.unlock()

  db_state = _fetch_db_state()
  if current != db_state return -ECANCELED
  return _update(current+new_delta);
}
```

A good candidate for the update() function could/should be `sfs::get_meta_buckets(get_store().db_conn)
      ->store_bucket()`, correct?

Reply:
I think this needs to be one layer deeper in the database update. Our Bucket code still keeps a shared copy of the database representation in a map, so this might be a bit more tricky with the database access and mutex around the map.

In essence I think we need something like 'update bucket set attrs = new attrs where attrs = cached attrs'
+ If that transaction didn't change anything, bail out and let the SAL layer update and retry.

Unfortunately I don't think can't say 'attrs = cached attrs', because both are ceph serialized blobs. A transaction that fetches, deserializes, compares and conditionally rolls back / commits may work.

Reply:
I've done a patch in this direction that seemed promising, but in the end I stumbled in an issue that IMO requires comments from upstream, so I've opened: https://tracker.ceph.com/issues/63560